File: app/src/main/java/ir/ghararemaghzha/game/adapters/ChatAdapter.java
package ir.ghararemaghzha.game.adapters;

import android.util.Base64;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.emoji.widget.EmojiTextView;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.textview.MaterialTextView;

import java.nio.charset.StandardCharsets;
import java.util.Objects;

import io.realm.OrderedRealmCollection;
import io.realm.Realm;
import io.realm.RealmRecyclerViewAdapter;
import io.realm.RealmResults;
import ir.ghararemaghzha.game.R;
import ir.ghararemaghzha.game.classes.DateConverter;
import ir.ghararemaghzha.game.classes.MySharedPreference;
import ir.ghararemaghzha.game.classes.Utils;
import ir.ghararemaghzha.game.data.RetrofitClient;
import ir.ghararemaghzha.game.models.MessageModel;
import ir.ghararemaghzha.game.models.TimeResponse;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ChatAdapter extends RealmRecyclerViewAdapter<MessageModel, RecyclerView.ViewHolder> {

    private FragmentActivity context;
    private final static int TYPE_ME = 1;
    private final static int TYPE_OTHER = 2;

    public ChatAdapter(FragmentActivity context, @Nullable OrderedRealmCollection<MessageModel> data, boolean autoUpdate) {
        super(data, autoUpdate);
        this.context = context;
        setHasStableIds(true);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public int getItemViewType(int position) {
        if (Objects.requireNonNull(getItem(position)).getSender().equals("1"))
            return TYPE_OTHER;
        else
            return TYPE_ME;

    }


    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == TYPE_ME)
            return new MeViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.row_chat_me, parent, false));
        else
            return new OtherViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.row_chat_other, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        final MessageModel model = getItem(position);
        if (model != null)
            try {
                if (getItemViewType(position) == TYPE_ME) {
                    MeViewHolder h = (MeViewHolder) holder;
                    byte[] data = Base64.decode(model.getMessage(), Base64.DEFAULT);
                    String text = new String(data, StandardCharsets.UTF_8);
                    h.message.setText( text);
                    String date = model.getDate();
                    DateConverter dateConverter = new DateConverter();

                    dateConverter.gregorianToPersian(Integer.parseInt(date.substring(0, 4)), Integer.parseInt(date.substring(5, 7)), Integer.parseInt(date.substring(8, 10)));
                    h.time.setText(date.substring(11, 16));
                    h.date.setText(dateConverter.getYear()+"/"+dateConverter.getMonth()+"/"+dateConverter.getDay());

                    switch (model.getStat()) {
                        case 0:
                            h.stat.setImageResource(R.drawable.vector_sending);
                            break;
                        case 1:
                            h.stat.setImageResource(R.drawable.vector_sent);
                            break;
                        case -1:
                            h.stat.setImageResource(R.drawable.vector_failed);
                            break;
                    }

                    h.stat.setOnClickListener(v -> {
                        if (model.getStat() == -1) {
                            h.stat.setImageResource(R.drawable.vector_sending);
                            byte[] data2 = model.getMessage().getBytes(StandardCharsets.UTF_8);
                            String body = Base64.encodeToString(data2, Base64.DEFAULT);

                            sendMessage(body, model.getMessageId(), position);
                        }
                    });


                } else {
                    OtherViewHolder h = (OtherViewHolder) holder;
                    byte[] data = Base64.decode(model.getMessage(), Base64.DEFAULT);
                    String text = new String(data, StandardCharsets.UTF_8);
                    h.message.setText( text);

                    String date = model.getDate();
                    DateConverter dateConverter = new DateConverter();

                    dateConverter.gregorianToPersian(Integer.parseInt(date.substring(0, 4)), Integer.parseInt(date.substring(5, 7)), Integer.parseInt(date.substring(8, 10)));
                    h.time.setText(date.substring(11, 16));
                    h.date.setText(dateConverter.getYear()+"/"+dateConverter.getMonth()+"/"+dateConverter.getDay());

                }


            } catch (Exception e) {
                e.printStackTrace();
            }
    }

    static class OtherViewHolder extends RecyclerView.ViewHolder {
        private MaterialTextView date,time;
        private EmojiTextView message;

        public OtherViewHolder(@NonNull View v) {
            super(v);
            message = v.findViewById(R.id.chat_row_message);
            date = v.findViewById(R.id.chat_row_date);
            time = v.findViewById(R.id.chat_row_time);
        }
    }

    static class MeViewHolder extends RecyclerView.ViewHolder {
        private MaterialTextView  date,time;
        private EmojiTextView message;
        private ImageView stat;

        public MeViewHolder(@NonNull View v) {
            super(v);
            message = v.findViewById(R.id.chat_row_message);
            date = v.findViewById(R.id.chat_row_date);
            stat = v.findViewById(R.id.chat_row_stat);
            time = v.findViewById(R.id.chat_row_time);

        }
    }

    private void sendMessage(String message, int key, int pos) {
        Realm db = Realm.getDefaultInstance();
        String number = MySharedPreference.getInstance(context).getNumber();
        String token = MySharedPreference.getInstance(context).getAccessToken();
        if (number.isEmpty() || token.isEmpty()) {
            Utils.logout(context,true);
            return;
        }
        RetrofitClient.getInstance().getApi()
                .sendMessage("Bearer " + token, number, message)
                .enqueue(new Callback<TimeResponse>() {
                    @Override
                    public void onResponse(@NonNull Call<TimeResponse> call, @NonNull Response<TimeResponse> response) {
                        if (response.isSuccessful() && response.body() != null && response.body().getResult().equals("success")) {
                            db.beginTransaction();
                            RealmResults<MessageModel> models = db.where(MessageModel.class).equalTo("messageId", key).findAll();
                            models.first().setStat(1);
                            db.commitTransaction();

                        } else {
                            db.beginTransaction();
                            RealmResults<MessageModel> models = db.where(MessageModel.class).equalTo("messageId", key).findAll();
                            models.first().setStat(-1);
                            db.commitTransaction();
                        }
                        notifyItemChanged(pos);

                    }

                    @Override
                    public void onFailure(@NonNull Call<TimeResponse> call, @NonNull Throwable t) {
                        db.beginTransaction();
                        RealmResults<MessageModel> models = db.where(MessageModel.class).equalTo("messageId", key).findAll();
                        models.first().setStat(-1);
                        db.commitTransaction();
                        notifyItemChanged(pos);

                    }
                });
    }

}

File: app/src/main/java/ir/ghararemaghzha/game/adapters/ChatAdapter.kt
package ir.ghararemaghzha.game.adapters

import android.app.Activity
import android.content.Context
import android.util.Base64
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.emoji.widget.EmojiTextView
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.textview.MaterialTextView
import io.realm.OrderedRealmCollection
import io.realm.Realm
import io.realm.RealmRecyclerViewAdapter
import io.realm.kotlin.where
import ir.ghararemaghzha.game.R
import ir.ghararemaghzha.game.classes.DateConverter
import ir.ghararemaghzha.game.classes.MySharedPreference
import ir.ghararemaghzha.game.classes.Utils
import ir.ghararemaghzha.game.data.ApiRepository
import ir.ghararemaghzha.game.data.NetworkApi
import ir.ghararemaghzha.game.data.RemoteDataSource
import ir.ghararemaghzha.game.data.Resource
import ir.ghararemaghzha.game.models.MessageModel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.nio.charset.StandardCharsets

class ChatAdapter(private val ctx: Context, data: OrderedRealmCollection<MessageModel>) : RealmRecyclerViewAdapter<MessageModel, RecyclerView.ViewHolder>(data, true) {
    private val typeMe = 1
    private val typeOther = 2

    init {
        setHasStableIds(true)
    }

    class MeViewHolder(v: View) : RecyclerView.ViewHolder(v) {
        val message: EmojiTextView = v.findViewById(R.id.chat_row_message)
        val date: MaterialTextView = v.findViewById(R.id.chat_row_date)
        val time: MaterialTextView = v.findViewById(R.id.chat_row_time)
        val stat: ImageView = v.findViewById(R.id.chat_row_stat)
    }

    class OtherViewHolder(v: View) : RecyclerView.ViewHolder(v) {
        val message: EmojiTextView = v.findViewById(R.id.chat_row_message)
        val date: MaterialTextView = v.findViewById(R.id.chat_row_date)
        val time: MaterialTextView = v.findViewById(R.id.chat_row_time)
    }

    override fun getItemId(position: Int): Long = getItem(position)?.messageId?.toLong() ?: 0

    override fun getItemViewType(position: Int): Int = if (getItem(position)?.sender == "1") typeOther else typeMe


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
            if (viewType == typeMe) MeViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.row_chat_me, parent, false))
            else
                OtherViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.row_chat_other, parent, false))


    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val model = getItem(position)
        if (model != null)
            if (getItemViewType(position) == typeMe) {
                val h = holder as MeViewHolder
                val data = Base64.decode(model.message, Base64.DEFAULT)
                val text = String(data, StandardCharsets.UTF_8)
                h.message.text = text
                val date = model.date
                val dateConverter = DateConverter()
                dateConverter.gregorianToPersian(date.substring(0, 4).toInt(), date.substring(5, 7).toInt(), date.substring(8, 10).toInt())
                h.time.text = date.substring(11, 16)
                h.date.text = "${dateConverter.year.toString().substring(2,4)}/${dateConverter.month}/${dateConverter.day}"
                when (model.stat) {
                    0 -> h.stat.setImageResource(R.drawable.vector_sending)
                    1 -> h.stat.setImageResource(R.drawable.vector_sent)
                    -1 -> h.stat.setImageResource(R.drawable.vector_failed)
                }
                h.stat.setOnClickListener {
                    if (model.stat == -1) {
                        h.stat.setImageResource(R.drawable.vector_sending)
                        val data2 = model.message.toByteArray(StandardCharsets.UTF_8)
                        val body = Base64.encodeToString(data2, Base64.DEFAULT)
                        CoroutineScope(Dispatchers.IO).launch {
                            sendMessage(body, model.messageId, position)
                        }
                    }
                }
            } else {
                val h = holder as OtherViewHolder
                val data = Base64.decode(model.message, Base64.DEFAULT)
                val text = String(data, StandardCharsets.UTF_8)
                h.message.text = text
                val date = model.date
                val dateConverter = DateConverter()
                dateConverter.gregorianToPersian(date.substring(0, 4).toInt(), date.substring(5, 7).toInt(), date.substring(8, 10).toInt())
                h.time.text = date.substring(11, 16)
                h.date.text = "${dateConverter.year.toString().substring(2,4)}/${dateConverter.month}/${dateConverter.day}"
            }

    }

    private suspend fun sendMessage(message: String, key: Int, pos: Int) {
        val number = MySharedPreference.getInstance(ctx).getNumber()
        val token = MySharedPreference.getInstance(ctx).getAccessToken()
        if (number.isEmpty() || token.isEmpty()) {
            Utils.logout(ctx as Activity, true)
            return
        }
        when (val res = ApiRepository(RemoteDataSource().getApi(NetworkApi::class.java)).sendMessage("Bearer $token", number, message)) {
            is Resource.Success -> {
                withContext(Dispatchers.Main) {
                    val db = Realm.getDefaultInstance()
                    if (res.value.result == "success") {
                        db.beginTransaction()
                        val model = db.where<MessageModel>().equalTo("messageId", key).findFirst()
                        model?.stat = 1
                        db.commitTransaction()
                    } else {
                        db.beginTransaction()
                        val models = db.where<MessageModel>().equalTo("messageId", key).findFirst()
                        models?.stat = -1
                        db.commitTransaction()
                    }
                    notifyItemChanged(pos)
                }
            }
            is Resource.Failure -> {
                if (res.isNetworkError) {
                    withContext(Dispatchers.Main) {
                        val db = Realm.getDefaultInstance()
                        db.beginTransaction()
                        val models = db.where<MessageModel>().equalTo("messageId", key).findFirst()
                        models?.stat = -1
                        db.commitTransaction()
                        notifyItemChanged(pos)
                    }
                } else if (res.errorCode == 401) {
                    withContext(Dispatchers.Main) {
                        Utils.logout(ctx as Activity, true)
                    }
                }
            }
        }
    }
}